python/tvm/contrib/cc.py

```diff
@@ -90,7 +90,8 @@ def get_target_triple():
 def cross_compiler(compile_func,
                    options=None,
                    output_format=None,
-                   get_target_triple=None):
+                   get_target_triple=None,
+                   add_files=None):
```

Comment:
this is a little bit of a weird place (compiler configuration time) to add extra files (I.e. invocation configuration time). why can't this be pushed to the invocation of fcompile?

Reply:
Indeed it is an API tradeoff. The main advantage of functor decorator style is that it is composable.

For example, we could pass use the following to attach minserver to the ndk compiler. 

```c++
mod.export_library("xyz", tvm.rpc.with_minrpc(cc.ndk))
```

Another reason is that in the case of minrpc, we need to attach a special property(need_system_lib) to tell the export_library to check for the `--system-lib` option.  

We might need a bit more thoughts to make the alternative API(as follows) work.
```c++
mod.export_library("xyz", cc.ndk, libs=[tvm.rpc.minrpc()])
```

Finally,  such functor decorator is easier to work with in the AutoTVM(we just have to pass the decorated version as build_func, instead of passing an additional argument)

Reply:
I am not very attached to the argument though. Pehaps changig to something like `default_libs` or `add_libs` would alleviate the concern?

Reply:
it seems like based on how this is used now, the problem is that we are trying to make export_library build a binary in a few small cases (but, which will become increasingly common as µTVM is more integrated). i think it's reasonable for export_library to control the compiler invocation up to the point of generating a .o. beyond that ldflags and additional statically-linked dependencies (which will have their own cflags and compiler toolchains) seem like a bit much of an API abuse. maybe this will be better addressed by the compilation and linking work i'm doing? if so, is there an intermediate hack we can do that avoids adding to this API (if you agree that's the right direction to go)?

Reply:
I think it is mainly a difference in API:

- A0  pushing more flags to export_library 
- A1 using functional style programming decorator, to decorate a existing compiler by adding ldflags implicitly. Pass in fcompile variants, or use fcompile->fcompile functors to decorate new compilers.

My take is that A1 is as powerful as A0, and the complexity of compiler customization goes to the decoration process.

We can certainly de-couple the APIs in A1. Right now compilation and linking are coupled, and a more powerful way to construct A1 might be as follows, and we can work toward that direction. 
```python
fcompile = attach_linker_opts(get_compiler())
```

Reply:
okay, I think de-coupling the APIs in A1 makes sense. if we do that though, then only autotvm (and other user scripts and any µTVM veneers) need to care about linker flags and linker scripts, correct? we can remove that complexity from `export_library`.

Reply:
I think your understanding is correct